### ISSUE TITLE
Question about get_prefix_rep

### ISSUE THREAD

**Issue by shivakanthsujit** [nakamotoo/dsrl_pi0#11]
Hello! 

I was looking at the real world experiments with openpi and was porting the changes from your fork of openpi to mine and noticed an issue with the `get_prefix_rep` function that you use.

You get the embedding like so, where `img_rep_pi0` is of shape (1, num_tokens, token_dim)

```py
img_rep_pi0, _ = agent_dp.get_prefix_rep(request_data)
img_rep_pi0 = img_rep_pi0[:, -1, :] # (1, 2048)
```
 And then you take the representation of the last token. However, the model transforms for $\pi$ pad the language tokens upto 200 tokens, even when passing in a single observation. So the last token is most likely a padding token, and we are taking the output token for a padding token, which I would understand to just be noise? Since the attention mask for that token is also False. So the noise actor and critic dont learn anything from having them in the input.

Ideally instead of just taking the last token, we should take the last non padding token index for image_rep. I do it like so

```py
        last_true_from_end = jnp.argmax(prefix_mask[:, ::-1], axis=1)  # 0-based from end
        last_valid_idx = prefix_mask.shape[1] - 1 - last_true_from_end
        # Handle case where all entries are False
        last_valid_idx = jnp.where(prefix_mask.any(1), last_valid_idx, -1)  # -1 if no valid tokens
        img_rep_pi0 = jnp.take_along_axis(hidden_state, last_valid_idx[:, None, None], axis=1)[:, 0, :]
```

Also when using the websocket client for the policy, passing the entire hidden state and kv_cache over the network can be a bottleneck (the kv cache is (2, 18, 1, 968, 1, 256), and is not necessary since we need only one output token. It would take a second or more for just the network passing which tanked my control frequency.  Instead this indexing can be done inside the get_prefix_rep function in openpi, making it take only about 50ms end to end over the network.
